MarkBind/markbind
# Move markdown-it parsing to /markbind

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] Documentation update
• [x] Other, please explain:

Resolves #969 - move attribute markdown parsing to /markbind, removing markdown-it as a dependency from vue-strap
Resolves #36 - standardise attribute naming for "header", "title", etc.

As a result,
Fixes #967 - some unexpected behaviours related to panel heading indexing
Combined with #966 ( closed ) - fix heading indexing for pure html headings, which is necessary for #969 

Requires https://github.com/MarkBind/vue-strap/pull/120 for the corresponding vue-strap template changes






**What is the rationale for this request?**

(1) To move all markdown parsing from the client to site building, which has benefits elaborated in #969 . Briefly,
- Better code cohesion & consistency ( where markdown parsing is done )
- Potentially faster client webpage loading ( due to not parsing markdown there )
- Smaller bundle size ( lack of markdown-it library from /vue-strap )

(2) Fix heading indexing for pure html headings ( originaly in #966 ). This is required for moving markdown parsing, as headings are now rendered in the dom as <hN> tags.

(3) Following the attribute parsing, attribute naming is conveniently standardised as well, and existing docs / srces / other usages updated. Deprecated attributes are still supported and documented. Components affected:
- dropdowns - "text" -> "header"
- boxes - "heading" -> "header"
- popovers - "title" -> "header"
- modals - "title" -> "header"

**What changes did you make? (Give an overview)**
(1) Introduce html heading id assigning identical to markdown-it-anchor for heading indexing, but for all types of headings. Consequently removed markdown-it-anchor plugin. Integrated this into `_parse` process.
(2) Introduced a generic per-component parser class housing all the new logic. Currently this only contains logic for parsing markdown in attributes. Integrated this into `_parse` process as well.
- It also houses all the attribute naming standardisation

**Provide some example code that this change will affect:**



The main changes are summarised in the `_parse` function on a high level.
```js
  // this handles the heading indexing
  if ((/^h[1-6]$/).test(element.name) && !element.attribs.id) {
    const textContent = utils.getTextContent(element);
    const slugifiedHeading = slugify(textContent, { decamelize: false });

    ... <some omitted id conflict resolution code> ...
  ... <old switch case> ...

  // this handles the attribute parsing
  self.componentParser.parseVueComponents(element);
```

**Is there anything you'd like reviewers to focus on?**
`ComponentParser` is quite unit - testable. This pr is getting a little large though - should it be done in a follow up instead?

**Testing instructions:**
`npm run test` will verify the dom changes to the test files, which have been checked.

**Proposed commit message: (wrap lines at 72 characters)**
Move markdown-it parsing to /markbind

Markdown passed in attributes are being rendered in
the client.

This results in a dependency on the markdown parsing
library ( markdown-it ) on the client bloating the resulting
bundle size for /vue-strap.
It also slows webpage loading, especially on older devices.
Moreover, /markbind already parses some amount of
markdown in attributes for other purposes when building
the site. This results in a less cohesive codebase.

Let's move all markdown-it parsing to /markbind from
/vue-strap, passing the rendered html into vue slots.
Additionally, let's add heading indexing for pure html
headings, which is required for heading indexing to
function correctly after this change.
